registry/distribution/repository.go

```diff
@@ -0,0 +1,674 @@
+func (r *Repository) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
+	return r.blobStore(target).Fetch(ctx, target)
+}
+
+// Push pushes the content, matching the expected descriptor.
+func (r *Repository) Push(ctx context.Context, expected ocispec.Descriptor, content io.Reader) error {
+	return r.blobStore(expected).Push(ctx, expected, content)
+}
+
+// Exists returns true if the described content exists.
+func (r *Repository) Exists(ctx context.Context, target ocispec.Descriptor) (bool, error) {
+	return r.blobStore(target).Exists(ctx, target)
+}
+
+// Delete removes the content identified by the descriptor.
+func (r *Repository) Delete(ctx context.Context, target ocispec.Descriptor) error {
+	return r.blobStore(target).Delete(ctx, target)
+}
+
+// Blobs provides access to the blob CAS only, which contains config blobs,
+// layers, and other generic blobs.
+func (r *Repository) Blobs() registry.BlobStore {
+	return &blobStore{repo: r}
+}
+
+// Manifests provides access to the manifest CAS only.
+func (r *Repository) Manifests() registry.BlobStore {
+	return &manifestStore{repo: r}
+}
+
+// Resolve resolves a reference to a manifest descriptor.
+func (r *Repository) Resolve(ctx context.Context, reference string) (ocispec.Descriptor, error) {
+	return r.Manifests().Resolve(ctx, reference)
+}
+
+// Tag tags a manifest descriptor with a reference string.
+func (r *Repository) Tag(ctx context.Context, desc ocispec.Descriptor, reference string) error {
+	ref, err := r.parseReference(reference)
+	if err != nil {
+		return err
+	}
+
+	rc, err := r.Manifests().Fetch(ctx, desc)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+
+	return r.push(ctx, desc, rc, ref.Reference)
+}
+
+// PushTag pushes the manifest with a reference tag.
+func (r *Repository) PushTag(ctx context.Context, expected ocispec.Descriptor, content io.Reader, reference string) error {
+	ref, err := r.parseReference(reference)
+	if err != nil {
+		return err
+	}
+	return r.push(ctx, expected, content, ref.Reference)
+}
+
+// push pushes the content, matching the expected descriptor.
+func (r *Repository) push(ctx context.Context, expected ocispec.Descriptor, content io.Reader, reference string) error {
+	url := fmt.Sprintf("%s/manifests/%s", r.endpoint(), reference)
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, url, content)
+	if err != nil {
+		return err
+	}
+	if req.GetBody != nil && req.ContentLength != expected.Size {
+		// short circuit a size mismatch for built-in types.
+		return fmt.Errorf("mismatch content length %d: expect %d", req.ContentLength, expected.Size)
+	}
+	req.ContentLength = expected.Size
+	req.Header.Set("Content-Type", expected.MediaType)
+
+	resp, err := r.Client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusCreated {
+		return parseErrorResponse(resp)
+	}
+	return verifyContentDigest(resp, expected.Digest)
```

Comment:
Given the recent CVE, I think we should verify the descriptor with `mediaType`, `digest` and `size` and not just digest.

Reply:
Can you share the CVE for detials since this code path is a `Push` path not a `Pull` path.

Reply:
Does put validate the response mediaType to the same that was PUT?

Reply:
The PUT response does not have a media type / `Content-Type`, and `Content-Length` is `0`.

See doc: https://github.com/distribution/distribution/blob/v2.7.1/docs/spec/api.md?plain=1#L1982-L1999

Reply:
In the OCI spec, even `Docker-Content-Digest` header is optional in the response.